### DISCUSSION TITLE
Proposal: Allow optional property initialization with fallback to the default value

### DISCUSSION THREAD

**Discussion by nrhoffmann** [dotnet/csharplang#1369] [Language Ideas]
### Proposal
Can we get this syntax `[Property Name] ?= [Some Value]`, when using the object initializer, to only override a property's default value if `[Some Value]` is not `null` or `default`?

### Example / Explanation
Currently, you can do this...
```c#
public class Connection
{
  public string Name { get; set; }
  public string Server { get; set; } = "127.0.0.1";
  public int Port { get; set; } = 80;
}
```

This is really nice, because it allows you to initialize a property with a default value. For example...
```c#

 var localhost = new Connection
 {
   Name = "localhost"   
 };

connections[localhost.Name] = localhost;
```

Suppose, that while refactoring my code I notice that I frequently make connections and add them to some dictionary that maps their name to themselves. To clean things up a bit I might do the following...
```c#
public Connection MapNewConnection(string name, string server = default, int port = default)
{
  var connection = new Connection
  {
    Name = name,
    Server = server,
    Port = port 
  };
  connections[connection.Name] = connection;
  return connection;
}
```

However, now I've lost the defaults set in `Connection`...
```c#
Console.WriteLine(MapNewConnection("localhost"));
// Connection { Name = localhost, Server = null, Port = 0 }
Console.WriteLine(MapNewConnection("google", "172.217.10.46"));
// Connection { Name = localhost, Server = "172.217.10.46", Port = 0 }
```

I am proposing the using the following syntax to solve this issue...
```c#
public Connection MapNewConnection(string name, string server = default, int port = default)
{
  var connection = new Connection
  {
    Name = name,
    Server ?= server,
    Port ?= port 
  };
  connections[connection.Name] = connection;
  return connection;
}
```

**Comment by vladd** [comment]
Why not just use `Server = server ?? default`?

**Comment by jnm2** [comment]
I like `??=` because it follows the existing patterns we have for converting an operator to an assignment shorthand operator, but `??` has always been about `null` and not `default`. Both of these things have been covered in previous issues https://github.com/dotnet/csharplang/issues/34 and https://github.com/dotnet/csharplang/issues/328.

**Comment by nrhoffmann** [comment]
@jnm2
Are they planning on making #34 work in object initializers?

What I mean is... 
According to #34 `x ??= y` would be syntactic sugar for `x = (y ?? x)`. However, in the above example `Server = (server ?? Server)` doesn’t compile.

My understanding was that `[Property Name] = [Some Value]` in an object initializer wasn't exactly a regular assignment.

Additionally, I would prefer `default` because properties are by default initialized to `default`. Thus, when a programmer writes ` public string Server { get; set; } = "127.0.0.1"`, he is conceptually overriding the default.

So I see it as "initialize the property to this value unless this value is the default value, because the programmer doesn't want that property to be initialized with that default value."

**Comment by YairHalberstadt** [comment]
`??=` has now shipped, but is not supported in object initializers. If this still interests you, I would suggest opening a new issue allowing it in object initializers.
